handler/src/main/java/io/netty/handler/ssl/ReleasableOpenSslContext.java

```diff
@@ -0,0 +1,789 @@
+                        switch (config.selectorFailureBehavior()) {
+                            case CHOOSE_MY_LAST_PROTOCOL:
+                            case NO_ADVERTISE:
+                                return new OpenSslDefaultApplicationProtocolNegotiator(
+                                        config);
+                            default:
+                                throw new UnsupportedOperationException(
+                                        new StringBuilder("OpenSSL provider does not support ")
+                                                .append(config.selectorFailureBehavior())
+                                                .append(" behavior").toString());
+                        }
+                    default:
+                        throw new UnsupportedOperationException(
+                                new StringBuilder("OpenSSL provider does not support ")
+                                        .append(config.selectedListenerFailureBehavior())
+                                        .append(" behavior").toString());
+                }
+            default:
+                throw new Error();
+        }
+    }
+
+    static boolean useExtendedTrustManager(X509TrustManager trustManager) {
+        return PlatformDependent.javaVersion() >= 7 && trustManager instanceof X509ExtendedTrustManager;
+    }
+
+    static boolean useExtendedKeyManager(X509KeyManager keyManager) {
+        return PlatformDependent.javaVersion() >= 7 && keyManager instanceof X509ExtendedKeyManager;
+    }
+
+    @Override
+    public final int refCnt() {
+        return refCnt;
+    }
+
+    @Override
+    public final ReferenceCounted retain() {
+        for (;;) {
+            int refCnt = this.refCnt;
+            if (refCnt == 0 || refCnt == Integer.MAX_VALUE) {
+                throw new IllegalReferenceCountException(refCnt, 1);
+            }
+            if (refCntUpdater.compareAndSet(this, refCnt, refCnt + 1)) {
+                break;
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public final ReferenceCounted retain(int increment) {
+        if (increment <= 0) {
+            throw new IllegalArgumentException("increment: " + increment + " (expected: > 0)");
+        }
+
+        for (;;) {
+            int refCnt = this.refCnt;
+            if (refCnt == 0 || refCnt > Integer.MAX_VALUE - increment) {
+                throw new IllegalReferenceCountException(refCnt, increment);
+            }
+            if (refCntUpdater.compareAndSet(this, refCnt, refCnt + increment)) {
+                break;
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public final ReferenceCounted touch() {
+        return touch(null);
+    }
+
+    @Override
+    public final ReferenceCounted touch(Object hint) {
+        if (leak != null) {
+            leak.record(hint);
+        }
+        return this;
+    }
+
+    @Override
+    public final boolean release() {
```

Comment:
I can do this ... extra object allocation shouldn't be a big deal ... if it is we can revert.